apis/projects/v1alpha1/variable_types.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+)
+
+// VariableType indicates the type of the GitLab CI variable.
+type VariableType string
+
+// List of variable type values.
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/project_level_variables.html
+const (
+	VariableTypeEnvVar VariableType = "env_var"
+	VariableTypeFile   VariableType = "file"
+)
+
+// VariableParameters define the desired state of a Gitlab CI Variable
+// https://docs.gitlab.com/ee/api/project_level_variables.html
+type VariableParameters struct {
+	// ProjectID is the ID of the project to create the variable on.
+	// +optional
+	// +immutable
+	ProjectID *int `json:"projectId,omitempty"`
+
+	// ProjectIDRef is a reference to a project to retrieve its projectId.
+	// +optional
+	// +immutable
+	ProjectIDRef *xpv1.Reference `json:"projectIdRef,omitempty"`
+
+	// ProjectIDSelector selects reference to a project to retrieve its projectId.
+	// +optional
+	ProjectIDSelector *xpv1.Selector `json:"projectIdSelector,omitempty"`
+
+	// Key for the variable.
+	// +kubebuilder:validation:Pattern:=^[a-zA-Z0-9\_]+$
+	// +kubebuilder:validation:MaxLength:=255
+	// +immutable
+	Key string `json:"key"`
+
+	// Value for the variable.
+	Value string `json:"value"`
+
+	// Masked enables or disables variable masking.
+	// +optional
+	Masked *bool `json:"masked,omitempty"`
+
+	// Protected enables or disables variable protection.
+	// +optional
+	Protected *bool `json:"protected,omitempty"`
+
+	// VariableType is the type of the variable.
+	// +kubebuilder:validation:Enum:=env_var;file
+	// +optional
+	VariableType *VariableType `json:"variableType,omitempty"`
+
+	// EnvironmentScope indicates the environment scope
+	// that this variable is applied to.
+	// +optional
+	EnvironmentScope *string `json:"environmentScope,omitempty"`
+}
+
+// VariableObservation represents a CI variable.
+type VariableObservation struct{}
```

Comment:
We can probably just omit this if it's unused. (Unless that makes angryjet think this type isn't a managed resource? I forget.)